File: lib/src/core/tag/tech/felica/felica.dart
part of core.tag;

/// The class provides access to `NFCFeliCaTag` API for iOS.
class Felica extends NFCTag {
  /// Constructs an instance with the given values for testing.
  ///
  /// The instances constructs by this way are not valid in
  /// the production environment.
  Felica._({
    required NFCTagPlatform delegate,
    required this.currentSystemCode,
    required this.currentIDm,
  }) : super._(delegate);

  /// Get an instance of `Felica` for the given tag.
  ///
  /// Returns null if the tag is not compatible with `Felica`.
  static Felica? _from({
    required NFCTagPlatform delegate,
  }) {
    if ( delegate.type.isNotFelica ) {
      return null;
    }

    final _data = Map<String, dynamic>.from(delegate.data);

    return Felica._(
      delegate: delegate,
      currentSystemCode: _data['currentSystemCode'],
      currentIDm: _data['currentIDm']);
  }

  /// The value from [NFCFeliCaTag#currentSystemCode] on iOS.
  final Uint8List currentSystemCode;

  /// The value from [NFCFeliCaTag#currentIDm] on iOS.
  final Uint8List currentIDm;

  /// Sends the `Polling` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#polling] API on iOS.
  Future<FelicaPollingResponse> polling({
    required Uint8List systemCode,
    required FelicaPollingRequestCode requestCode,
    required FelicaPollingTimeSlot timeSlot,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'polling',
          'systemCode': systemCode,
          'requestCode': requestCode.value,
          'timeSlot': timeSlot.value,
        })
        .then(
          (value) => FelicaPollingResponse.fromMap(
            Map<String, dynamic>.from(value!)));

  /// Sends the `Request Response` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#requestResponse] API on iOS.
  Future<int> requestResponse() =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'requestResponse',
        })
        .then((value) => value!);

  /// Sends the `Request System Code` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#requestSystemCode] API on iOS.
  Future<List<Uint8List>> requestSystemCode() =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'requestSystemCode',
        })
        .then(
          (value) => List<Uint8List>.from(value!));

  /// Sends the `Request Service` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#requestService] API on iOS.
  Future<List<Uint8List>> requestService({
    required List<Uint8List> nodeCodeList,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'requestService',
          'nodeCodeList': nodeCodeList,
        })
        .then(
          (value) => List<Uint8List>.from(value!));

  /// Sends the `Request Service V2` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#requestServiceV2] API on iOS.
  Future<FelicaRequestServiceV2Response> requestServiceV2({
    required List<Uint8List> nodeCodeList,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'requestServiceV2',
          'nodeCodeList': nodeCodeList,
        })
        .then(
          (value) => FelicaRequestServiceV2Response.fromMap(
            Map<String, dynamic>.from(value!)));

  /// Sends the `Read Without Encryption` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#readWithoutEncryption] API on iOS.
  Future<FelicaReadWithoutEncryptionResponse> readWithoutEncryption({
    required List<Uint8List> serviceCodeList,
    required List<Uint8List> blockList,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'readWithoutEncryption',
          'serviceCodeList': serviceCodeList,
          'blockList': blockList,
        })
        .then(
          (value) => FelicaReadWithoutEncryptionResponse.fromMap(
            Map<String, dynamic>.from(value!)));

  /// Sends the `Write Without Encryption` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#writeWithoutEncryption] API on iOS.
  Future<FelicaStatusFlag> writeWithoutEncryption({
    required List<Uint8List> serviceCodeList,
    required List<Uint8List> blockList,
    required List<Uint8List> blockData,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'writeWithoutEncryption',
          'serviceCodeList': serviceCodeList,
          'blockList': blockList,
          'blockData': blockData,
        })
        .then(
          (value) => FelicaStatusFlag.fromMap(
            Map<String, dynamic>.from(value!)));

  /// Sends the `Request Specification Version` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#requestSpecificationVersion] API on iOS.
  Future<FelicaRequestSpecificationVersionResponse>
      requestSpecificationVersion() =>
    channel.invokeMethod(
      "Felica", {
        'handle': handle,
        'method': 'requestSpecificationVersionResponse',
      })
      .then(
        (value) => FelicaRequestSpecificationVersionResponse.fromMap(
          Map<String, dynamic>.from(value!)));

  /// Sends the `Reset Mode` command to the tag.
  ///
  /// This uses [NFCFeliCaTag#resetMode] API on iOS.
  Future<FelicaStatusFlag> resetMode() =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'resetMode',
        })
        .then(
          (value) => FelicaStatusFlag.fromMap(
            Map<String, dynamic>.from(value!)));

  /// Sends the `Felica` command packet data to the tag.
  ///
  /// This uses [NFCFeliCaTag#sendFeliCaCommand] API on iOS.
  Future<List<int>> sendCommand({
    required Uint8List commandPacket,
  }) =>
      channel.invokeMethod(
        "Felica", {
          'handle': handle,
          'method': 'sendCommand',
          'commandPacket': commandPacket,
        })
        .then(
          (value) => Int8List.fromList(value!).toList());

  @override
  String toString() =>
      "[Felica] "
      "currentSystemCode: ${currentSystemCode.toHexString()}, "
      "currentIDm: ${currentIDm.toHexString()}";
}

File: ios/Classes/Felica.swift
import CoreNFC
import Flutter

internal class Felica: NSObject {
    private let tagManager: TagManager

    public init(_ tagManager: TagManager) {
        self.tagManager = tagManager
    }

    @available(iOS 13.0, *)
    public func handle(_ arguments: [String: Any?], result: @escaping FlutterResult) {
        switch arguments["method"] {
        case "polling": polling(arguments, result: result)
        case "requestResponse": requestResponse(arguments, result: result)
        case "requestSystemCode": requestSystemCode(arguments, result: result)
        case "requestService": requestService(arguments, result: result)
        case "requestServiceV2": requestServiceV2(arguments, result: result)
        case "readWithoutEncryption": readWithoutEncryption(arguments, result: result)
        case "writeWithoutEncryption": writeWithoutEncryption(arguments, result: result)
        case "requestSpecificationVersion": requestSpecificationVersion(arguments, result: result)
        case "resetMode": resetMode(arguments, result: result)
        case "sendCommand": sendCommand(arguments, result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    @available(iOS 13.0, *)
    private func Polling(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let systemCode = (arguments["systemCode"] as! FlutterStandardTypedData).data
            let requestCode = PollingRequestCode(
                rawValue: arguments["requestCode"] as! Int)!
            let timeSlot = PollingTimeSlot(
                rawValue: arguments["timeSlot"] as! Int)!

            tag.polling(systemCode: systemCode, requestCode: requestCode, timeSlot: timeSlot) { manufacturerParameter, requestData, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "manufacturerParameter": manufacturerParameter,
                        "requestData": requestData,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func RequestResponse(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            tag.requestResponse { mode, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result(mode)
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func RequestSystemCode(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            tag.requestSystemCode { systemCodeList, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result(systemCodeList)
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func RequestService(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let nodeCodeList = (arguments["nodeCodeList"] as! [FlutterStandardTypedData]).map { $0.data }

            tag.requestService(nodeCodeList: nodeCodeList) { nodeKeyVersionList, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result(nodeKeyVersionList)
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func RequestServiceV2(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let nodeCodeList = (arguments["nodeCodeList"] as! [FlutterStandardTypedData]).map { $0.data }

            tag.requestServiceV2(nodeCodeList: nodeCodeList) { statusFlag1, statusFlag2, encryptionIdentifier, nodeKeyVersionListAes, nodeKeyVersionListDes, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "statusFlag1": statusFlag1,
                        "statusFlag2": statusFlag2,
                        "encryptionIdentifier": encryptionIdentifier.rawValue,
                        "nodeKeyVersionListAes": nodeKeyVersionListAes,
                        "nodeKeyVersionListDes": nodeKeyVersionListDes,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func ReadWithoutEncryption(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let serviceCodeList = (arguments["serviceCodeList"] as! [FlutterStandardTypedData]).map { $0.data }
            let blockList = (arguments["blockList"] as! [FlutterStandardTypedData]).map { $0.data }

            tag.readWithoutEncryption(serviceCodeList: serviceCodeList, blockList: blockList) { statusFlag1, statusFlag2, blockData, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "statusFlag1": statusFlag1,
                        "statusFlag2": statusFlag2,
                        "blockData": blockData,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func WriteWithoutEncryption(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let serviceCodeList = (arguments["serviceCodeList"] as! [FlutterStandardTypedData]).map { $0.data }
            let blockList = (arguments["blockList"] as! [FlutterStandardTypedData]).map { $0.data }
            let blockData = (arguments["blockData"] as! [FlutterStandardTypedData]).map { $0.data }

            tag.writeWithoutEncryption(serviceCodeList: serviceCodeList, blockList: blockList, blockData: blockData) { statusFlag1, statusFlag2, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "statusFlag1": statusFlag1,
                        "statusFlag2": statusFlag2,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func RequestSpecificationVersion(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            tag.requestSpecificationVersion { statusFlag1, statusFlag2, basicVersion, optionVersion, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "statusFlag1": statusFlag1,
                        "statusFlag2": statusFlag2,
                        "basicVersion": basicVersion,
                        "optionVersion": optionVersion,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func ResetMode(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            tag.resetMode { statusFlag1, statusFlag2, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result([
                        "statusFlag1": statusFlag1,
                        "statusFlag2": statusFlag2,
                    ])
                }
            }
        }
    }

    @available(iOS 13.0, *)
    private func sendCommand(_ arguments: [String : Any?], result: @escaping FlutterResult) {
        self.tagManager.tagHandler(NFCFeliCaTag.self, arguments, result) { tag in
            let commandPacket = (arguments["commandPacket"] as! FlutterStandardTypedData).data

            tag.sendFeliCaCommand(commandPacket: commandPacket) { data, error in
                if let error = error {
                    result(getFlutterError(error))
                } else {
                    result(data)
                }
            }
        }
    }
}